Big ol' refactor of deploy logic

msheiny opened the pull request:
Fixes #23, #18 so far -- will have to re-evaluate some others. What does this PR change? 

So - currently in `master`, the role logic involves remoting into a prod server and installing all of the pre-reqs and making subsequent changes. This has caused problems in production, especially on a live site and when pip requirements change in weird ways. So the changes introduced in this PR add two big things:

* All of the build logic is conducted locally under docker containers -- including pulling site code, virtualenv building, node installation. This means less dependencies installed on the server.
* We add parallel set of gunicorn services that toggle back and forth between which one is active (basically pointing the reverse-proxy, in our case `nginx` to one service or the other at the conclusion of every deploy).

A few things to keep in mind:
* This PR breaks backwards compatibility in a big way. So any previous deploys we did using this role will basically have to have the servers rebuilt. This is fine because we are using ansible so those servers aren't snowflakes :)
* The deploys will take slightly longer because whenever there is a code change we are now building pip requirements and node logic from scratch instead of building upon what was already on the server. Ultimately this is the right call -- and I tried to make the role smart enough to only do that when there is a code change (idempotency seems to be working fine).

Right right so you can test with:
* Install all the reqs under `devops/requirements.txt` into a virtualenv
* Run `make`

I'm throwing up this PR a little early to get some feedback though since its such a breaking change. I suspect there are a few dangling default variables that arent be used anymore. I'll try to clean this up as well.

msheiny committed:
Migrate to testing logic to molecule v2

This provides a number of benefits -- the biggest of which is we can ditch a lot
of the custom logic that was in the `devops` folder.

msheiny committed:
Refactor role to build locally

This is a pretty substantial commit - the result of a big rebase during a
long and boring trial-test-repeat phase. Anywhoo - this commit dramatically
changes the logic of the role. Where as before the "compilation" of the
virtualenv and node logic was being done on the server, now that logic is being
completed client-side under a docker container. Because of this, this commit
severely breaks backwards compatibility with playbooks using this role before
it.

One important thing to note is that server-wise there are now two parallel
gunicorn services that are stood up with them being cycled back and forth with
an nginx snippet. The reasoning is that we can severely reduce weird
side-effects during the time when code is being pushed up to a live server.

msheiny committed:
Disable CircleCI build logic during transition

The new role build logic involves docker containers which are doing local file
mounts. Unfortunately this wont play nice with CircleCI so lets remove these
files until we can get this working under CI again. Until then, manual running
is the name of the game :(

msheiny committed:
Update readme and make commands

msheiny committed:
Ensure virtualenv container is destroyed

Not sure whether this should be also done in the role itself... that'll be a
commit for another day.

msheiny committed:
Address yamllint and ansible-lint issues

Was hitting a bunch of warnings regarding line-limit which I dont see as a big
deal here so I set to disable errors on those (instead they are now warnings).

msheiny committed:
Fix for npm global + no command

Found a scenario where the install would fail when global npm packages were
needed BUT an npm command was not needed to run against the code base :|

msheiny committed:
Refactor CircleCI to use ephemeral VMs

This will save us some time from having to refactor the docker logic to work
with a docker remote executor. Hopefully CircleCI doesnt start charging in the
near future *fingers crossed*

msheiny committed:
sync - ensure perms are writable by the gcorn user

Previously was keeping the perms from the docker container.. this didnt trigger
an error in CI but it did in staging. Hooray for staging!

msheiny committed:
Upstream docker_container module needed for role

When I attempted to use latest role commit in an outside playbook this
dependency came up. Its basically the docker_container module utilized in
ansible 2.4 with a `working_dir` argument. Once 2.4 is released this can be
purged from the role.

msheiny committed:
Ensure permissions get laid out correctly

Synchronize module wasnt cutting it unfortunately. So had to follow up with a
recursive chown smackdown

conorsch commented:
Finally sitting down to review this monster. Molecule v2 integration looks great!

conorsch commented on README.md:
```diff
@@ -132,29 +151,34 @@ django_stack_force_refresh: false
 If you want to experiment with your own site repo, you'll either have to start
 folding this role into your deployment playbook and experiment there or start
-making local edits to `devops/playbook.yml` accordingly.
+making local edits to `molecule/cis/playbook.yml` accordingly.
```
Nit: `s/cis/ci/`

conorsch commented on defaults/main.yml:
```diff
@@ -45,12 +53,13 @@ django_stack_es_host_url: http://localhost:9200
 
 #npm settings
-django_stack_npm_install_cmd: npm install
+django_stack_npm_install_cmd: "npm install; npm run start"
```
Shouldn't the chaining logic here use `&&` rather than `;`?

msheiny replied:
i honestly dont think it matters -- ive had a script bomb out before here and it just kills the entire container and the play stops with an error. im hesitant to change it though -- its running in `ash` , not `bash` under a alpine linux container

conorsch commented on tasks/npm.yml:
```diff
@@ -1,48 +1,14 @@
-  changed_when: false
+- name: Establish some npm specific vars for current play
+  set_fact:
+    npm_touch: "touch /status/npm.done"
```
This tempfile logic is difficult to understand. Some comments would be warranted. Near as I can figure, this task is referencing a static location incorrectly, when it should be using `"{{ tmp_dir }}/npm.done" instead. More explicit var names would help here, e.g. `django_stack_local_tmp_dir`.

conorsch replied:
Ah, I see: the `/status` location is hardcoded because it's mounted via volumes. Take this a nit for comments, then. I'll add!

conorsch committed:
Consolidates NPM logic into single task list

The indirection between the `tasks/npm.yml` and `tasks/npm-code.yml`
made it difficult to understand the control flow used by the role.
Combined the two task lists, but preserved the `block` conditional to
skip the logic from `npm-code` as before.

conorsch committed:
Fixes typo in README for Molecule playbook path

conorsch committed:
Cleans up local containers after role run

The role was previously allowing the helper containers created on the
Ansible controller, for preparing the build dependencies, to dangle
after the role completed. Added a final step to stop them and remove the
associated volumes.

In the process, I renamed pf_tracker_node -> django_stack_node for
consistency with the django_stack_venv container name.

conorsch commented:
Ingenious use of tooling to get more reliable deployments, @msheiny. You had to jump through a lot of hoops you wouldn't've had to if were deploying with containers directly, but, well, _one day_ :wistful_sigh:.

Addressed my own nits during review, and reorged the task flow a bit. I also added a cleanup task to remove the helper containers on the Ansible controller after the role run completes.

conorsch reviewed (approved):
In general I have concerns about the liberal use of `set_fact`, but the role has become quite dynamic, so there's not really a better way. Top job. 

During review I did _not_ rename all the `tmp_dir` vars, but I was certainly tempted to. I'll just leave this here: https://a-nickels-worth.blogspot.com/2016/04/a-guide-to-naming-variables.html

msheiny commented:
I concede it is a confusing mess of variables -- i swear ill doing a second sweep of var cleanup when I add more tests in #26  :pinky: . Thanks for your kind words and sweet review skills ❤️
